### PR TITLE
[MM-16174] Migrate Team.SaveMember to Sync by default #11125

### LINKED ISSUE #11125: Migrate "Team.SaveMember" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-16174)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `SaveMember` in the `Team` store.

The expected way to implement it is, go to the `Team` store implementation in the `store/sqlstore/` directory, modify the method `SaveMember` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### PR DESCRIPTION
**Pull request by Inconnu08**
#### Summary
Migrate `Team.SaveMember` to sync by default, removing the `return.Do` wrapper.

#### Ticket Link
Fixes #11125

### CODE REVIEW

**Comment by jespino** [comment]
@Inconnu08 there are still not migrated files, look at the jenkins output.

**Comment by Inconnu08** [comment]
@hanzei, I don't think I understood why this test is failing on Jenkins so I'm waiting for you guys to respond actually.😐

**Comment by jespino** [comment]
@Inconnu08 @hanzei Yes, I want to investigate it, but I didn't find the time. I'll try to do it right now.

**Comment by jespino** [comment]
@Inconnu08 some build problems here, do you can take a look into it?

**Comment by Inconnu08** [comment]
Roger that

File: Makefile [hunk 1]
```diff
@@ -330,7 +330,7 @@ i18n-extract: ## Extract strings for translation from the source code
 	$(GOPATH)/bin/mmgotool i18n extract
 
 store-mocks: ## Creates mock files.
-	env GO111MODULE=off go get -u github.com/vektra/mockery/...
+	env GO111MODULE=off go get -u github.com/vektra/mockery/cmd/mockery
```
Comment (grundleborg) [inline Makefile]: What's the reason for this change? It doesn't seem related to the rest of the PR.

File: app/team.go [hunk 2]
```diff
@@ -508,11 +508,12 @@ func (a *App) joinUserToTeam(team *model.Team, user *model.User) (*model.TeamMem
 	rtm, err := a.Srv.Store.Team().GetMember(team.Id, user.Id)
 	if err != nil {
 		// Membership appears to be missing. Lets try to add.
-		tmr := <-a.Srv.Store.Team().SaveMember(tm, *a.Config().TeamSettings.MaxUsersPerTeam)
-		if tmr.Err != nil {
-			return nil, false, tmr.Err
+		var tmr *model.TeamMember
+		tmr, err = a.Srv.Store.Team().SaveMember(tm, *a.Config().TeamSettings.MaxUsersPerTeam)
```
Comment (grundleborg) [inline app/team.go]: Can this be reduced to 1 line by just doing `tmr, err := ...`?

File: store/storetest/team_store.go [hunk 3]
```diff
@@ -770,54 +771,58 @@ func testTeamMembers(t *testing.T, ss store.Store) {
...
 
-	store.Must(ss.Team().SaveMember(m2, -1))
-	store.Must(ss.Team().SaveMember(m3, -1))
+	_, err := ss.Team().SaveMember(m2, -1)
+	require.Nil(t, err)
+	_, err = ss.Team().SaveMember(m3, -1)
+	require.Nil(t, err)
 
-	if ms, err := ss.Team().GetMembers(teamId1, 0, 100, nil); err != nil {
+	var ms []*model.TeamMember
```
Comment (grundleborg) [inline store/storetest/team_store.go]: Why has this been added and the assignments changed below?

File: store/storetest/team_store.go [hunk 4]
```diff
@@ -770,54 +771,58 @@ func testTeamMembers(t *testing.T, ss store.Store) {
...
+	if ms, err = ss.Team().GetMembers(teamId1, 0, 100, nil); err != nil {
```
Comment (grundleborg) [inline store/storetest/team_store.go]: You can simplify these if/else checks to just:
```
ss.Team().GetMembers(teamId1, 0, 100, nil)
require.Nil(t, err)
require.Len(t, ms, 2)
```
rather than needing the if conditions. If a `require` fails, the test case won't continue anyway. 0/5 on whether you make this change though - up to you.

File: app/team.go (hunk 2 above)
Comment (jespino) [reply app/team.go]: Probably there will generate a "shadowing" error, because the `err` variable is created in the new scope, and shadows the upper scope `err` variable.

File: Makefile (hunk 1 above)
Comment (jespino) [reply Makefile]: Done

File: store/storetest/team_store.go (hunk 3 above)
Comment (jespino) [reply store/storetest/team_store.go]: same shadowing variable problem, but with your next comment it gets simplified.

Comment (jespino) [reply store/storetest/team_store.go]: Fixed

File: store/storetest/team_store.go (hunk 4 above)
Comment (jespino) [reply store/storetest/team_store.go]: Fixed

**Comment by jespino** [comment]
I took over this PR to help @Inconnu08 finishing it

File: store/storetest/team_store.go [hunk 5]
```diff
@@ -770,54 +771,55 @@ func testTeamMembers(t *testing.T, ss store.Store) {
 	m2 := &model.TeamMember{TeamId: teamId1, UserId: model.NewId()}
 	m3 := &model.TeamMember{TeamId: teamId2, UserId: model.NewId()}
 
-	if r1 := <-ss.Team().SaveMember(m1, -1); r1.Err != nil {
-		t.Fatal(r1.Err)
+	if _, err := ss.Team().SaveMember(m1, -1); err != nil {
```
Comment (mgdelacroix) [inline store/storetest/team_store.go]: nit: `require.Nil(t ,err)` could be used here
